Jenkinsfile

```diff
@@ -154,13 +162,56 @@ node {
       if (params.BUILD_GRAVITY_APP) {
         withEnv(MAKE_ENV) {
           sh 'make export'
-          archiveArtifacts "build/application.tar"
         }
       } else {
         echo 'skipped build gravity app'
       }
     }
+
+    stage('upload application tarball to S3') {
+      if (isProtectedBranch(env.TAG) && params.PUBLISH_APP_PACKAGE) {
+        withCredentials([usernamePassword(credentialsId: "${AWS_CREDENTIALS}", usernameVariable: 'AWS_ACCESS_KEY_ID', passwordVariable: 'AWS_SECRET_ACCESS_KEY')]) {
+          def artifactName = "application.tar"
+          def s3AppName = "stolon-app"
+          if (params.BUILD_GRAVITY_HELM_APP) {
+            artifactName = "helm-application.tar"
+            s3AppName = "stolon-app-helm"
+          }
+          def s3Url = "s3://${S3_UPLOAD_PATH}/${s3AppName}:${APP_VERSION}.tar"
+          withEnv(MAKE_ENV + ["S3_URL=${s3Url}", "artifactName=${artifactName}"]) {
+            sh 'aws s3 cp --only-show-errors build/${artifactName} ${S3_URL}'
+          }
+        }
+      }
+    }
+  }
+}
+
+def isProtectedBranch(branchOrTagName) {
+  // check if tag or branch empty
+  if (!branchOrTagName?.trim()) {
+    return false
+  }
+
+  String[] protectedBranches = ['master', 'support/.*']
+  def match = false
+
+  protectedBranches.any { protectedBranch ->
+    if (branchOrTagName == "${protectedBranch}") {
+      match = true
+      return true
+    }
+    def status = sh(script: "git branch --all --contains=${branchOrTagName} | grep '[*[:space:]]*remotes/origin/${protectedBranch}\$'", returnStatus: true)
+    if (status == 0) {
+      match = true
+      return true
+    }
+  }
+
+  if (match) {
+    return true
   }
+  return false
```

Comment:
`any` ends early if it finds an element that matches the condition, so this code can be
```
return protectedBranches.any { protectedBranch ->
  if (branchOrTagName == "${protectedBranch}") return true
  def status = sh(script: "git branch --all --contains=${branchOrTagName} | grep '[*[:space:]]*remotes/origin/${protectedBranch}\$'", returnStatus: true)
  return status == 0
}